WalletWasabi/EventSourcing/InMemoryEventRepository.cs

```diff
@@ -0,0 +1,216 @@
+				IndexNewAggregateId(aggregateType, aggregateId);
+			}
+			return Task.CompletedTask;
+		}
+
+		public Task<IReadOnlyList<WrappedEvent>> ListEventsAsync(
+			string aggregateType,
+			string aggregateId,
+			long afterSequenceId = 0,
+			int? limit = null)
+		{
+			if (_aggregatesEventsBatches.TryGetValue(aggregateType, out var aggregateEventsBatches) &&
+				aggregateEventsBatches.TryGetValue(aggregateId, out var value))
+			{
+				var result = value.Events;
+				if (afterSequenceId > 0)
+				{
+					var foundIndex = result.BinarySearch(
+						new WrappedEvent(afterSequenceId),
+						Comparer<WrappedEvent>.Create((a, b) => a.SequenceId.CompareTo(b.SequenceId)));
+					if (foundIndex < 0)
+					{
+						foundIndex = ~foundIndex;
+					}
+					else
+					{
+						foundIndex++;
+					}
+					result = result.RemoveRange(0, foundIndex);
+				}
+				if (limit.HasValue && limit.Value < result.Count)
+				{
+					result = result.RemoveRange(limit.Value, result.Count - limit.Value);
+				}
+				return Task.FromResult((IReadOnlyList<WrappedEvent>)result);
+			}
+			return Task.FromResult(EmptyResult);
+		}
+
+		public Task<IReadOnlyList<string>> ListAggregateIdsAsync(
+			string aggregateType,
+			string? afterAggregateId = null,
+			int? limit = null)
+		{
+			if (_aggregatesIds.TryGetValue(aggregateType, out var tuple))
+			{
+				var ids = tuple.Ids;
+				var foundIndex = 0;
+				if (afterAggregateId != null)
+				{
+					foundIndex = ids.IndexOf(afterAggregateId);
+					if (foundIndex < 0)
+					{
+						foundIndex = ~foundIndex;
+					}
+					else
+					{
+						foundIndex++;
+					}
+				}
+				var result = new List<string>();
+				var afterLastIndex = limit.HasValue ?
+					Math.Min(foundIndex + limit.Value, ids.Count) :
+					ids.Count;
+				for (var i = foundIndex; i < afterLastIndex; i++)
+				{
+					result.Add(ids[i]);
+				}
+				return Task.FromResult((IReadOnlyList<string>)result.AsReadOnly());
+			}
+			return Task.FromResult(EmptyIds);
+		}
+
+		private void IndexNewAggregateId(string aggregateType, string id)
+		{
+			var tailIndex = 0L;
+			ImmutableSortedSet<string> aggregateIds;
+			ImmutableSortedSet<string> newAggregateIds;
+			var liveLockLimit = 10000;
+			do
+			{
+				if (liveLockLimit-- <= 0)
+				{
+					throw new ApplicationException("Live lock detected.");
+				}
+				(tailIndex, aggregateIds) = _aggregatesIds.GetOrAdd(aggregateType, _ => new(0, ImmutableSortedSet<string>.Empty));
+				newAggregateIds = aggregateIds.Add(id);
+			}
+			while (!_aggregatesIds.TryUpdate(
+				key: aggregateType,
+				newValue: (tailIndex + 1, newAggregateIds),
+				comparisonValue: (tailIndex, aggregateIds)));
+		}
+
+		// Helper for parallel critical section testing in DEBUG build only.
+		[Conditional("DEBUG")]
+		protected virtual void Validated()
+		{
+		}
+
+		// Helper for parallel critical section testing in DEBUG build only.
+		[Conditional("DEBUG")]
+		protected virtual void Conflicted()
+		{
+		}
```

Comment:
how come? IMO they are helpful for understanding the structure of the critical section, and what the invariants are, even if they weren't useful for testing purposes

Reply:
I would find it valuable to get rid of `[Conditional("DEBUG")]` attribute. The main reason being that devs test using DEBUG configuration and end users are on RELEASE configuration and imho we want to keep the difference between what we test and what end users use to be pretty minimal. One can always circumevent this using an interface/virtual method/mock etc. to properly test code. 

It's hard for me to propose anything here because the method is empty. I don't think we should add things saying "it will be useful in the future". It's better to add functionality as it is needed and when it's properly tested.

Reply:
but it's used and is properly tested?

Reply:
but it's used and is properly tested?

Reply:
can't be tested without it. Testing is important. I made it so it doesn't make it to the production build.

Reply:
it is virtual so it is overridden and therefore it is used even if empty

Reply:
> can't be tested without it. 

This is very suspicious to me because there are other approaches.

> Testing is important.

Agreed.

But then why not simply have https://github.com/kiminuo/WalletWasabi/tree/feature/2021-12-06-no-conditional to make the tests work regardless of DEBUG/RELEASE configuration?

Reply:
Let's be concrete. There are two different issues. One is the conditional compilation. Of course it doesn't has to be conditional. But there is a good reason for it. The method is not meant to be in production code. The method is there only as a hook to enable testing. Actually making it this way is the best for readability and prevention of future problems.

The other issue is the existence of the hook methods at all. I need those for testing. There is not other way of hooking into the middle of some method.

Reply:
> 

what is the link suppose to show? It's probably not working 🤷‍♂️

Reply:
> Let's be concrete. There are two different issues. One is the conditional compilation. Of course it doesn't has to be conditional. But there is a good reason for it.

And this is where our opinions differs. What you say makes sense but cons are bigger for me than the pros:

* We release the app using RELEASE configuration. Unit tests are currently run using DEBUG configuration in Azure Pipelines. In general, when you write some tests that work only in DEBUG configuration, you cannot be sure that the application behaves in the same way in the RELEASE configuration. For this reason alone, I find it dangerous. For a toy project, I wouldn't mind. For a project involving money, it's best to program in a defensive way.
    * A few months ago, I have debugged some bug that was caused by use of `[Conditional("Debug")]` and it's very painful because in that configuration you don't have typically logs, breakpoints work only to a degree in my IDE, etc.
* I have not found any `[Conditional("Debug")]` or `#if DEBUG` occurrences in WW tests, so a more persuassive argument from your side would be: Currently, you use in your tests X and Y (or nothing similar) but my conditional thingy is better because ... but you just keep repeating that "it's for tests", "can't be tested without", etc. This is simply not true.
* Honestly, the only thing that `[Conditional("DEBUG")]` gives you is that you skip a functional call which is very cheap thing to do (and that might actually be optimized away entirely). So when you simply remove those `[Conditional("DEBUG")]` attributes, you can run tests in the RELEASE configuration (that [branch](https://github.com/kiminuo/WalletWasabi/tree/feature/2021-12-06-no-conditional)). You might lose a tiny bit of performance but gain the confidence that what you release behaves correctly (much more important for me).

> The other issue is the existence of the hook methods at all. I need those for testing. There is not other way of hooking into the middle of some method.

I'm saying remove the `[Conditional("DEBUG")]` and all is good.

Reply:
> what is the link suppose to show?

It shows that one should not ...

> [..] It's probably not working 🤷‍♂️

... make conclusions before saying anything. :)